### PR TITLE
AirfRANS: gc=0.5 + T_max=10 extended — prevent ep205 divergence

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

gc=0.5 at 4L/256d achieved the best AirfRANS result ever (**0.00277**, PR #2774, your own run) but catastrophically diverged at ep205 — same T_max=5 instability that destroys gc=1.0 at ep208. The divergence is caused by T_max=5's aggressive cosine LR peaks (high LR every 5 epochs) accumulating instability over ~200 epochs until a gradient spike overwhelms clipping.

**Hypothesis:** T_max=10 (slower cycling) reduces LR peak frequency and amplitude, preventing the catastrophic ep205 divergence. If stable, gc=0.5 + T_max=10 should continue improving past ep205 and push below 0.002.

## Instructions

No code changes needed. Same winning config but with `--cosine-t-max 10`:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent roy \
  --wandb-name "roy/airfrans-4L256d-gc05-tmax10" \
  --wandb-group "airfrans-gc05-stability"
```

**Key things to report:**
1. Does the model survive past ep205? What are grad norms at ep200-210?
2. Best val_primary/surface_mse and at which epoch
3. How does convergence speed compare to T_max=5 (your previous run 0pt769m4)?
4. Trough envelope every 20 epochs

## Baseline

- **Current best:** val_primary/surface_mse = **0.00277** (PR #2774, YOUR RUN — gc=0.5 + T_max=5)
- **W&B run:** 0pt769m4 (221 epochs, diverged at ep205)
- **Multiple sub-0.004 troughs confirmed:** e77=0.00356, e183=0.00322, e204=0.00308
- **Reproduce (winning config):** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT roy:

## Results

### Key Metrics

| Metric | T_max=10 (this run) | T_max=5 baseline (0pt769m4) |
|--------|--------------------|-----------------------------|
| **Best val_primary/surface_mse** | **0.002086** (ep139) | 0.00277 (ep204) |
| Divergence epoch | **ep141** | ep205 |
| Total epochs | 220 (180-min timeout) | 221 (diverged ep205) |
| Useful epochs (before divergence) | 140 | 204 |

**Best = 0.002086 at ep139 — 24.7% below baseline (0.00277)** but diverged catastrophically 2 epochs later.

### Trough Envelope (cycle-end minima)

| Epoch | val_primary/surface_mse | grad_norm_mean |
|-------|------------------------|----------------|
| 20 | 0.07354 | 20.65 |
| 36 | 0.00969 | 10.97 |
| 59 | 0.00681 | 8.23 |
| 70 | 0.00378 | 7.07 |
| 100 | 0.00675 | 5.21 |
| 114 | 0.00355 | 4.50 |
| 125 | 0.00348 | 4.25 |
| 128 | 0.00282 | 4.23 |
| **139** | **0.00209** | 4.08 |
| 141 | 0.23091 (DIVERGED) | 232.78 |

### Divergence Timeline (ep139-147)

| Epoch | surface_mse | grad_norm_mean |
|-------|-------------|----------------|
| 139 | 0.002086 | 4.08 |
| 140 | 0.009684 | 4.70 |
| 141 | 0.230911 | 232.78 |
| 142 | 0.038155 | 624.54 |
| 143 | 0.502367 | 1,674.56 |
| 144 | 0.358326 | 3,908.98 |
| 145 | 0.194989 | 12,926.63 |
| 146 | 0.310550 | 57,991.35 |
| 147 | 0.254206 | inf |

Grad norms went from 4.08 → inf in 8 epochs. After ep147, the model was permanently stuck at ~0.255-0.273 with inf grads for the remaining 73 epochs.

### Convergence Speed Comparison vs T_max=5

T_max=10 reached sub-0.003 faster (ep128 = 0.00282 vs T_max=5 first sub-0.003 at ~ep183), but the trough envelope was less stable — larger oscillations between troughs and peaks (peaks hit 0.041-0.045 vs T_max=5 peaks ~0.02-0.03).

### What Happened

**Hypothesis REJECTED.** T_max=10 did not prevent divergence — it accelerated it by 64 epochs (ep141 vs ep205). The slower LR cycling actually made things worse:

1. **Faster trough descent** — T_max=10 pushed the model deeper into low-loss regions faster (0.00209 at ep139 vs 0.00277 at ep204), but this seems to have destabilized the model.
2. **Larger peak oscillations** — The longer cycle means LR stays high for longer periods, creating more violent swings (peaks of 0.041-0.045 vs ~0.02-0.03 for T_max=5).
3. **Same failure mode** — The divergence pattern is identical: a sudden grad norm explosion from stable (~4) to inf within ~8 epochs. This suggests the root cause is NOT LR peak frequency but rather an intrinsic instability in the model/loss landscape at this depth that accumulates regardless of cycling speed.
4. **The deeper the trough, the more unstable** — Both runs diverged shortly after hitting their best values, suggesting the model reaches a region of the loss landscape where it becomes critically unstable.

### Experiment Details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent roy --wandb-name "roy/airfrans-4L256d-gc05-tmax10" --wandb-group "airfrans-gc05-stability"`
- **W&B run:** sb75ib6q
- **Peak memory:** ~48.5 GB (GPU 0)
- **Duration:** 180 min (timeout), 220 epochs

### Suggested Follow-ups

1. **Lower LR** — The instability may stem from lr=7e-4 being too aggressive at 4L/256d. Try lr=3.5e-4 or lr=5e-4 with gc=0.5 and T_max=5.
2. **Stronger gradient clipping** — gc=0.5 wasn't enough to contain the explosion. Try gc=0.25 or gc=0.1.
3. **LR warmdown / decay** — Instead of fixed cosine cycling, use a decaying cosine schedule (reduce max LR over time) to tame late-training instability.
4. **Early stopping with checkpoint** — Since both runs hit their best right before divergence, an early-stopping mechanism based on gradient norm thresholds could capture the best checkpoint before collapse.
5. **Weight decay increase** — Higher weight decay (e.g., 5e-2) could regularize the model away from the unstable region.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2818 because ADVISOR: Closing — your T_max=10 result of 0.002086 does not beat the 0.001479 baseline. This falsifies the hypothesis that T_max=10 helps with divergence — it actually made things worse. Valuable negative result, roy. We now know T_max=5 is the sweet spot..
